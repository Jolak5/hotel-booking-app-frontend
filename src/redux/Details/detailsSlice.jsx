import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import { getLocalStorage } from '../../helpers/localStorage';

const initialState = {
  hoteldetails: [],
  iLoading: false,
  error: null,
};

const url = 'https://hotel-booking-7djb.onrender.com/hotels/';

export const fetchHotel = createAsyncThunk('details/fetchhotels', async (id, thunkAPI) => {
  try {
    const token = getLocalStorage('token');
    const response = await fetch(`${url}${id}`, {
      headers: {
        Authorization: `Bearer ${token}`,
      },
    });
    const data = await response.json();
    return data;
  } catch (error) {
    return thunkAPI.rejectWithValue(error.response.data);
  }
});

const detailsSlice = createSlice({
  name: 'details',
  initialState,
  reducers: {},
  extraReducers: {
    [fetchHotel.pending]: (state) => ({ ...state, isLoading: true }),
    [fetchHotel.fulfilled]: (state, action) => ({
      ...state,
      isLoading: false,
      hoteldetails: action.payload,
    }),
    [fetchHotel.rejected]: (state, action) => ({
      ...state,
      isLoading: false,
      error: action.payload,
    }),

  },
});

export default detailsSlice.reducer;
