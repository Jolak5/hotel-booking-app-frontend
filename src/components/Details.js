import { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useParams } from 'react-router-dom';
import { fetchHotel } from '../redux/Details/detailsSlice';
import styles from './Details.module.css';

const Details = () => {
  const { id } = useParams();
  const { isLoading, hoteldetails } = useSelector((state) => state.details);
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(fetchHotel(id));
  }, [dispatch, id]);

  if (isLoading) {
    return (
      <p>Loading...</p>
    );
  }
  return (
    <div className={styles.details}>
      <div className={styles.divimg}>
        {hoteldetails.image && <img className={styles.detailsimg} src={hoteldetails.image.url} alt={hoteldetails.name} />}
      </div>
      <div className={styles.divdetails}>
        <h3>{hoteldetails.name}</h3>
        <table className={styles.detailstable}>
          <tr className={styles.divtr}>
            <td> Hotel price</td>
            <td> {hoteldetails.price} </td>
          </tr>
          <tr className={styles.divtr}>
            <td> Duration </td>
            <td> {hoteldetails.duration} </td>
          </tr>
        </table>
        <button className={styles.reserve}> Reserve </button>
      </div>
    </div>
  );
};

export default Details;
