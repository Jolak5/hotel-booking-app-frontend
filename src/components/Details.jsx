import { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { FaGreaterThan } from 'react-icons/fa';
import { useParams, Link } from 'react-router-dom';
import { fetchHotel } from '../redux/Details/detailsSlice';
import styles from '../styles/Details.module.css';

const Details = () => {
  const { id } = useParams();
  const { isLoading, hoteldetails } = useSelector((state) => state.details);
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(fetchHotel(id));
  }, [dispatch, id]);

  if (isLoading) {
    return (
      <p>Loading...</p>
    );
  }
  return (
    <div className={styles.details}>
      {hoteldetails.image
        && <img className={styles.detimg} src={hoteldetails.image.url} alt={hoteldetails.name} />}
      <div className={styles.hoteltextdetails}>
        <div className={styles.detailer}>
          <h2 className={styles.hoteldetailsname}>{hoteldetails.name}</h2>
          <p className={styles.hoteldetailsdescription}>{hoteldetails.description}</p>
          <div className={styles.hotelpricecontainer}>
            <p className={styles.textheading}>Price Per Night</p>
            <span className="price-value">
              $
              {hoteldetails.price}
            </span>
          </div>
          <div className={styles.durationcontainer}>
            <p className={styles.textheading}>Duration</p>
            <span className="duration-value">
              {hoteldetails.duration}
              days
            </span>
          </div>
        </div>
        <div className={styles.links}>
          <Link to="/home" className={styles.morehotelslink}>
            Discover More Hotel Rooms
            <FaGreaterThan />
          </Link>
          <Link to="/reservations/new" className={styles.reserve}>
            Reserve
          </Link>
        </div>
      </div>
    </div>
  );
};

export default Details;
