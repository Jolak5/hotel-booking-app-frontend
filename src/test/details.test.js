import React from 'react';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import { Routes, Route, MemoryRouter } from 'react-router-dom';

import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import Details from '../components/Details';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

describe('Details', () => {
  const initialState = {
    details: {
      fetched: true,
      isLoading: false,
      hoteldetails: {
        id: 2,
        name: 'Hotel 2',
        description: 'This is a description for Hotel 2.',
        duration: 7,
        price: '139.13',
        image: {
          url: 'http://localhost:3000/rails/active_storage/blobs/redirect/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBCdz09IiwiZXhwIjpudWxsLCJwdXIiOiJibG9iX2lkIn19--16d773ef08a1a413ca00a961f9b8696c29001642/image2.jpg',
        },
      },
    },
  };

  const store = mockStore(initialState);
  const hotelId = '2';

  test('renders without crashing', () => {
    const { getByText } = render(
      <Provider store={store}>
        <MemoryRouter initialEntries={[`/${hotelId}`]}>
          <Routes>
            <Route path="/:id" element={<Details />} />
          </Routes>
        </MemoryRouter>
      </Provider>,
    );
    expect(getByText('Hotel 2')).toBeInTheDocument();
  });

  test('matches snapshot', () => {
    const tree = renderer.create(
      <Provider store={store}>
        <MemoryRouter initialEntries={[`/${hotelId}`]}>
          <Routes>
            <Route path="/:id" element={<Details />} />
          </Routes>
        </MemoryRouter>
      </Provider>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
