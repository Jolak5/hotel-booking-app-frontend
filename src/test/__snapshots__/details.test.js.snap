// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Details matches snapshot 1`] = `
<div
  className="details"
>
  <img
    alt="Hotel 2"
    className="detimg"
    src="http://localhost:3000/rails/active_storage/blobs/redirect/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBCdz09IiwiZXhwIjpudWxsLCJwdXIiOiJibG9iX2lkIn19--16d773ef08a1a413ca00a961f9b8696c29001642/image2.jpg"
  />
  <div
    className="hoteltextdetails"
  >
    <div
      className="detailer"
    >
      <h2
        className="hoteldetailsname"
      >
        Hotel 2
      </h2>
      <p
        className="hoteldetailsdescription"
      >
        This is a description for Hotel 2.
      </p>
      <div
        className="hotelpricecontainer"
      >
        <p
          className="textheading"
        >
          Price Per Night
        </p>
        <span
          className="price-value"
        >
          $
          139.13
        </span>
      </div>
      <div
        className="durationcontainer"
      >
        <p
          className="textheading"
        >
          Duration
        </p>
        <span
          className="duration-value"
        >
          7
          days
        </span>
      </div>
    </div>
    <div
      className="links"
    >
      <a
        className="morehotelslink"
        href="/home"
        onClick={[Function]}
      >
        Discover More Hotel Rooms
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 384 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M365.52 209.85L59.22 67.01c-16.06-7.49-35.15-.54-42.64 15.52L3.01 111.61c-7.49 16.06-.54 35.15 15.52 42.64L236.96 256.1 18.49 357.99C2.47 365.46-4.46 384.5 3.01 400.52l13.52 29C24 445.54 43.04 452.47 59.06 445l306.47-142.91a32.003 32.003 0 0 0 18.48-29v-34.23c-.01-12.45-7.21-23.76-18.49-29.01z"
          />
        </svg>
      </a>
      <button
        className="reserve"
        onClick={[Function]}
        type="submit"
      >
        Reserve
      </button>
    </div>
  </div>
</div>
`;
